WIP feat(views): make it easier to work with layout filters

hypeJunction opened the pull request:
Make it easier for plugins to extend layout filters without having to override views.
Update groups/all resource to use updated filters

hypeJunction committed:
feat(navigation): adds API to control content layout tabs

Make it easier for plugins to extend layout filters without having to
override views via elgg_get_filter_tabs().
Update groups/all resource to use updated filters.

hypeJunction commented:
Thinking out loud here. This will be a BC hell, but I think we could make it work.

mrclay commented on mod/groups/views/default/resources/elements/groups/alpha.php:
```diff
@@ -0,0 +1,15 @@
+<?php
```
I think we should reserve resource views to mostly complete pages. Why not just "groups/elements/alpha" etc.?

hypeJunction replied:
Sure, I had that there, but then thought it would be closer to resources. But, yes, I agree.
As a general rule, I want to move all elgg_list_entities() and elgg_view_form() calls into their own views, so that we don't have to touch resource views unnecessarily.

mrclay replied:
> move all elgg_list_entities() and elgg_view_form() calls into their own views

👍 👍

mrclay commented on mod/groups/lib/groups.php:
```diff
@@ -101,13 +102,58 @@ function groups_prepare_form_vars($group = null) {
+		),
+	);
+
+	if (elgg_is_active_plugin('discussions')) {
```
Belongs in discussions, right? Or is this an attempt at BC?

hypeJunction replied:
Yes, would move there eventually, if we can agree on a general roadmap.

mrclay commented:
I like it.

hypeJunction commented:
@jdalsem @jeabakker Could you take a quick look, as it will affect plugins like group_tools? If you are ok with the direction, I will figure out a BC way to make it work. Me likes myself a challenge.

jdalsem commented:
Looks good to me, however i would move the tabs into a menu... easier to manipulate... group tools will probably not be affected as it overrules both the sort menu view and the resources...

hypeJunction committed:
feat(menus): elgg_view_menu() now accepts an array of menu items

elgg_view_menu() now accepts an array of default menu items to add to the
unprepared menu bypassing menu registers. This is useful for rendering
multiple instances of the same menu on the page, e.g. filter

hypeJunction committed:
feat(discussions): break down discussion listing views

Adds resource subviews to render discussion lists

hypeJunction commented:
I think this is BC now.

jdalsem commented:
why not change the tabs registration to a menu hook?

hypeJunction commented:
You can already use the `menu:filter` hook

jdalsem commented:
having a groups_sort_menu to trigger a hook ('filter_tabs') to get tabs, to register menu items... seems a bit overkill... why not just register the menu items in a menu register hook?... that would mean not use the fiter_tabs hook for menu registration...

do we than even need the filter_tabs hook?

hypeJunction commented:
The point here is to allow passing an array of tabs to `elgg_view_layout()`.

hypeJunction commented:
@mrclay Thoughts on this? I am working with members plugin at the moment, so could just rewire the whole thing.

mrclay commented on mod/groups/start.php:
```diff
@@ -257,8 +257,36 @@ function groups_page_handler($page) {
 	$vars = [];
 	switch ($page[0]) {
+		case 'all' :
+			$plugin_root = __DIR__;
```
This all goes away in 3.0 so it's cool.

mrclay commented:
Something is resulting in `navigation/menu/default` receiving an array for `$vars['class']`. Let's just add handling for that.

mrclay commented:
LGTM. Would this make changing the default tab easier?

hypeJunction commented:
> Would this make changing the default tab easier?

You could hook into `filter_tabs` and reassign priorities.

hypeJunction committed:
feat(members): now uses views to render lists, updates tab registration

Updates tab registration to use elgg_get_filter_tabs().
Moves list rendering to views.

hypeJunction commented:
I had a change of hearts when I started updating the members plugin. There is now a more generic `elgg_get_filter_tabs()`. I am leaning toward deprecating the use of magic `filter` in content views and updating all core plugins to set the filter explicitly.

hypeJunction commented on engine/lib/navigation.php:
```diff
@@ -641,6 +641,75 @@ function _elgg_nav_public_pages($hook_name, $entity_type, $return_value, $params
+	}
+
+	uasort($tabs, function ($a, $b) {
+		return elgg_extract('priority', $a, 500) - elgg_extract('priority', $b, 500);
```
@mrclay How do I get this to work properly across PHP 5 and 7? It currenty reverses the array of priorities are not set.

mrclay replied:
You mean returning 0 here causes a different sort between 5 and 7?

mrclay replied:
If so you might look at what hooks/events do: https://github.com/Elgg/Elgg/blob/2.x/engine/classes/Elgg/HooksRegistrationService.php#L203

mrclay replied:
I'm starting to see @jdalsem's point. This is starting to feel like menu code. Not having to learn another API would be nice.

hypeJunction replied:
uasort() seems to have a mind of its own. 
I do see his point but I doubt it's going to work, at least not in a BC way. If you or him have the energy to try, please do - I already need another holiday. 
Sorting here is used to later determine the default (fallback) tab. I guess we can add a function to get a det a default tab from a set of tabs that triggers a hook as well.

hypeJunction commented:
I am going to let it sit until I have an aha moment. I don't think using menus directly is going to work.

mrclay commented:
Yeah, pause this until we can answer a few Qs: 1. If BC were no concern, why is the menu system unsuitable for tabs? 2. If a menu would work, do we want to create a new API in 2.3 just to replace it in 3.0? Keeping this new API in 3.0 might be worth it to solve today's problems in 2.3 (are they a huge pain point?)

mrclay commented:
As for u(a)sort, comparison operations can't return 0. Before sort, number the tabs by array order and in the comparison function fallback to that ordering if priority is the same.

mrclay commented:
👆 I'll send you a PR

hypeJunction commented:
Current problem is that some execution paths never reach elgg_view_menu(). We need to purge default tabs and use the menu hook, but we would still need to inject resource name/context into layout vars, which won't be BC.
My pain point is the ability to extend tabs in any give context, e.g. I want tl add a tab to blog/edit to add other forms, .e.g seo meta tags - extending the form woukd be an option in 2.3, but that wouldn't solve problems with non-form resources, e.g. adding tabs for privacy, tools, notofications to tge settings form. 
I can live without this in 2.x but this is a real use case that needs attention. Just imagine a layout with a sidebar and a page menu. That makes navigation impossible.

hypeJunction commented:
Sorry for typos. Without* sidebar I meant.

hypeJunction commented:
I guess we should merge the menu commit, I think it will be handy. I will extract it into a separate PR

hypeJunction commented:
Menu commit: https://github.com/Elgg/Elgg/pull/10213

hypeJunction commented:
A less controversial version: #10256

hypeJunction commented:
I am going to close this for now. I am thinking of adding a new layout type that would allow us to do this in a BC way, but I need to ponder with the idea.
